### PR TITLE
Implement message trailer parsing API

### PR DESCRIPTION
**Pull request by charliesome**
This pull request adds a `git_message_trailer` function that parses trailers out of commit messages. This function takes the message as a NUL-terminated C string and a callback. This callback is called once per trailer found. The callback can return non-zero at any time to stop trailer processing and return immediately.

The logic for trailer parsing is mostly taken directly from git core's `trailer.c`, so this will behave the same way as git around all the complexities of trailer parsing such as conflicts, patches, trailer continuations, etc.

I've also added a bunch of tests with all the edge cases I can think of.

cc @brianmario

### CODE REVIEW

**Comment by ethomson** [comment]
@pks-t would your parser state machine be helpful here?

File: src/trailer.c [hunk 1]
```diff
@@ -0,0 +1,415 @@
...
+
+static const char *const git_generated_prefixes[] = {
+	"Signed-off-by: ",
+	"(cherry picked from commit ",
+	NULL
+};
+
+static const char *const separators = ":";
+
+static inline int is_blank_line(const char *str)
```
Comment (ethomson) [inline src/trailer.c]: `inline` needs to be `GIT_INLINE`

File: src/trailer.c [hunk 2]
```diff
@@ -0,0 +1,415 @@
...
+	i = len - 2;
+
+	for (; i >= 0; i--) {
+		if (buf[i] == '\n')
+			return i + 1;
+	}
+	return 0;
+}
+
+static bool starts_with(const char *str, const char *prefix)
```
Comment (ethomson) [inline src/trailer.c]: This looks like we could instead use our existing `git__prefixcmp`.

File: src/trailer.c (hunk 1 above)
Comment (charliesome) [reply src/trailer.c]: I might just drop the `inline` entirely. It's a small static function so it'll probably end up being inlined automatically anyway.

**Comment by brianmario** [comment]
I have the rugged branch going over at https://github.com/libgit2/rugged/compare/brianmario/trailers-api but it's failing because the cibuild script appears to be forcing the `libgit2` submodule at `master`. The tests pass locally, so once this is shipped I'll make sure things are all good to go on the rugged side and get a pull open there as well.

**Comment by brianmario** [comment]
@ethomson Hi and happy new year! 👋  

How are you feeling about this now? Would love to get cranking on the rugged side so we can start using it 😄 

Pinging @carlosmn as well just in case one of you is back online before the other 😉

**Comment by ethomson** [comment]
Hey @brianmario - I hadn't looked at it - I didn't realize y'all were ready for review, since it's not compiling on Windows.  Could you take a look at the errors?

**Comment by brianmario** [comment]
The link to join the slack team isn't working for some reason, but I'm unsure what the error syntax is trying to communicate exactly.

> ..\..\tests\message\trailer.c(45): error C2059: syntax error : '{'

Would lead me to believe line 45 of `trailer.c` is where the issue is, but that and the lines around it are a multi-line code comment. Of which the `{` character is nowhere to be found. I imagine it's something to do with some "newer" (>C89 ?) C syntax that isn't compatible with older versions of VS (2013 in this case). But I'm not seeing it without help from these errors ;)

**Comment by tiennou** [comment]
I think inline struct initialization is at least C90, and it's used in the test file.

**Comment by pks-t** [comment]
@ethomson: I assume you mean the parser API instead of the state machine. And I think this patch would indeed benefit from it, as once again this here duplicates logic which is already provided by `git_parser`. So I would heavily encourage you (@brianmario) to use that one instead. If there are shortcomings in the existing API, we can surely extend it.

File: src/trailer.c [hunk 3]
```diff
@@ -0,0 +1,415 @@
...
+ * a Linking Exception. For full terms see the included COPYING file.
+ */
+#include "common.h"
+#include "git2/message.h"
+
+#include <stddef.h>
+#include <string.h>
+#include <ctype.h>
+
+static const char comment_line_char = '#';
```
Comment (pks-t) [inline src/trailer.c]: I feel like this should probably be a define instead of a const variable.

File: src/trailer.c [hunk 4]
```diff
@@ -0,0 +1,415 @@
...
+				continue;
+		} else if (isspace(bol[0]))
+			possible_continuation_lines++;
+		else {
+			non_trailer_lines++;
+			non_trailer_lines += possible_continuation_lines;
+			possible_continuation_lines = 0;
+		}
+continue_outer_loop:
+		;
```
Comment (pks-t) [inline src/trailer.c]: I'd prefer to have a continue here such that one can easily see that we're iterating the loop.

File: src/trailer.c [hunk 5]
```diff
@@ -0,0 +1,415 @@
...
+			if (recognized_prefix &&
+			    trailer_lines * 3 >= non_trailer_lines)
+				return next_line(bol) - buf;
+			else if (trailer_lines && !non_trailer_lines)
+				return next_line(bol) - buf;
+			return len;
+		}
+		only_spaces = 0;
+
+		for (p = git_generated_prefixes; *p; p++) {
```
Comment (pks-t) [inline src/trailer.c]: Why do we treat git generated prefixes differently? Trailer lines can be whatever information the user wants them to be, so in my opinion there is no need to check for git prefixes but instead to check for the expected trailer line format.

File: src/trailer.c [hunk 6]
```diff
@@ -0,0 +1,415 @@
...
+
+static char *find_trailer(const char *message, size_t* len)
+{
+	size_t patch_start = find_patch_start(message);
+	size_t trailer_end = find_trailer_end(message, patch_start);
+	size_t trailer_start = find_trailer_start(message, trailer_end);
+
+	size_t trailer_len = trailer_end - trailer_start;
+
+	char *buffer = git__malloc(trailer_len + 1);
```
Comment (pks-t) [inline src/trailer.c]: Not exactly sure, but is this even allowed in C90 wrt variable declaration after statement?

File: src/trailer.c [hunk 7]
```diff
@@ -0,0 +1,403 @@
...
+	S_VALUE = 4,
+	S_VALUE_NL = 5,
+	S_VALUE_END = 6,
+	S_IGNORE = 7,
+};
+
+#define NEXT(st) { state = (st); ptr++; continue; }
+#define GOTO(st) { state = (st); continue; }
+
+int git_message_trailers(const char *message, git_message_trailer_cb cb, void *payload)
```
Comment (pks-t) [inline src/trailer.c]: The whole loop could use a state machine as we do for other parts (e.g. patch parsing, see patch_parse.c, parse_header_git())

File: tests/message/trailer.c [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+		"\n"
+		"Signed-off-by: foo@bar.com\n"
+		"Signed-off-by: someone@else.com\n"
+	,
+		(struct trailer[]){
+			{ "Signed-off-by", "foo@bar.com" },
+			{ "Signed-off-by", "someone@else.com" },
+			{ NULL, NULL },
+		}
+	);
```
Comment (pks-t) [inline tests/message/trailer.c]: I think this would be a bit more readable if you didn't declare the arrays as part of the macro but instead as variables first and then pass them into the assert_trailers. This would also get rid of the cast.

File: include/git2/message.h [hunk 9]
```diff
@@ -38,6 +38,29 @@ GIT_BEGIN_DECL
...
+ * @param message The message to be parsed
+ *
+ * @param cb The callback to call for each trailer found in the message. The
+ *     key and value arguments are pointers to NUL-terminated C strings. These
+ *     pointers are only guaranteed to be valid until the callback returns.
+ *     User code must make a copy of this data should it need to be retained
+ *
+ * @param payload Pointer to callback data (optional)
+ *
+ * @return 0 on success, or non-zero callback return value.
```
Comment (pks-t) [inline include/git2/message.h]: Small nit: we usually don't use empty lines between @-lines.

File: src/trailer.c [hunk 10]
```diff
@@ -0,0 +1,403 @@
...
+
+#include <stddef.h>
+#include <string.h>
+#include <ctype.h>
+
+static const char comment_line_char = '#';
+
+static const char *const git_generated_prefixes[] = {
+	"Signed-off-by: ",
+	"(cherry picked from commit ",
```
Comment (pks-t) [inline src/trailer.c]: Is this really a trailer line? I thought the definition of a trailer line is that it is of the form "key: value"

File: src/trailer.c [hunk 11]
```diff
@@ -0,0 +1,403 @@
...
+	NULL
+};
+
+static const char *const separators = ":";
```
Comment (pks-t) [inline src/trailer.c]: Same as above

File: tests/message/trailer.c (hunk 8 above)
Comment (pks-t) [reply tests/message/trailer.c]: Btw: this also causes MSVC 2010 builds to fail.

**Comment by ethomson** [comment]
I'm luke warm on shipping it as-is since often we never _actually_ get around to iterating on things.  However I'm sympathetic to your process of including libgit2/rugged updates where you literally ship `master`.  (VSTS ships a fork so that we can merge things before they land here in `master`.)

So if you need this to unblock you on a feature, I'm sympathetic.  And we can take a quick review pass.  **BUT** I would want a firm guarantee that we will continue to refactor this, and promptly.

Does that seem reasonable?

File: src/trailer.c (hunk 10 above)
Comment (brianmario) [reply src/trailer.c]: I think this is one of those git-core-specific intricacies I was talking about. I think technically a trailer can be in just about any configurable format (there are git-config settings we'll need to implement support for here, eventually). The HTTP-style `Key: value` stuff is just the default.

It looks like this stuff was added in git-core in https://github.com/git/git/commit/146245063e286d5a38b146bd5a38da958bd3957b - forgot there was also a heuristic used to determine what a trailer block is...

File: src/trailer.c (hunk 5 above)
Comment (brianmario) [reply src/trailer.c]: My [comment below](https://github.com/libgit2/libgit2/pull/4451#discussion_r159573210) has some context around this but so you don't have to skip around (that comment is shown as outdated now for some reason) - https://github.com/git/git/commit/146245063e286d5a38b146bd5a38da958bd3957b has more context on this. tldr; a trailer block is determined using a heuristic of "regular" trailers and git-created trailers.

Comment (charliesome) [reply src/trailer.c]: This logic is taken directly from Git. It seems pretty funky to me, but the goal here is to be compatible with Git so I've brought over this git generated prefix logic as-is.

**Comment by brianmario** [comment]
@ethomson yeah, I think that's fair 😃 

I've made some of the requested changes, and either wasn't sure about the others or am intentionally holding off (like rewriting the parser) for now.

File: src/trailer.c (hunk 6 above)
Comment (charliesome) [reply src/trailer.c]: I believe so - we're still in variable declaration territory at this point since the only expressions preceding this are initializers for other declarations.

File: tests/message/trailer.c [hunk 12]
```diff
@@ -0,0 +1,200 @@
...
+
+	cl_assert_equal_i(0, rc);
+}
+
+void test_message_trailer__simple(void)
+{
+	struct trailer trailers[3];
+
+	trailers[0].key   = "Signed-off-by";
+	trailers[0].value = "foo@bar.com";
```
Comment (charliesome) [inline tests/message/trailer.c]: I think we can still use braced array initializers in ANSI C. Something like this should do the trick:

```c
struct trailer trailers[] = {
  { "Signed-off-by", "foo@bar.com" },
  /* ... */
  { NULL, NULL },
};
```

File: src/trailer.c (hunk 4 above)
Comment (charliesome) [reply src/trailer.c]: Just to clarify, replace this lone semicolon with a `continue;`?

File: tests/message/trailer.c (hunk 12 above)
Comment (pks-t) [reply tests/message/trailer.c]: Yep :+1:

File: src/trailer.c (hunk 5 above)
Comment (pks-t) [reply src/trailer.c]: Yeah, I definitly agree that we should strive to follow upstream-logic as close as possible. I really don't like all these workarounds and heuristics being put into so many places just because the initial design wasn't flexible enough. Nothing we can do about that... thanks though for the pointers!

File: src/trailer.c (hunk 4 above)
Comment (pks-t) [reply src/trailer.c]: At least that's what I'd think is more readable. The loop is quite large, so you can easily forget that you're inside of one. Though now that I come to think of it the label already says 'continue_outer_loop', I didn't spot that one before. So this may also stay like it currently is

File: src/trailer.c (hunk 6 above)
Comment (pks-t) [reply src/trailer.c]: I think I saw a compiler once treating intermingled function calls, assignments and newlines inside of the declaration block spitting an error. I bet it was some MSVC-style thing, but I cannot quite recall it.

File: src/trailer.c (hunk 5 above)
Comment (brianmario) [reply src/trailer.c]: We may have not been very clear in the OP or throughout this discussion, but this isn't "workarounds and heuristics" of the initial design - it's [literally the upstream code itself](https://github.com/git/git/blob/master/trailer.c#L861). Which by my understanding is the source of truth for how any of this is supposed to work anyway, right?

The design of the code may not match the standards of libgit2, but if we (libgit2) are to handle trailers exactly the same way git does, we unfortunately need to also take the heuristics in to account whenever we refactor this code anyway.

For now at least, literally using the upstream code seems like the safest (at least in terms of compatibility) way to go.

**Comment by brianmario** [comment]
Is there a way to force a rebuild on appveyor? The current failures were due to an external request failing.

File: src/trailer.c (hunk 5 above)
Comment (pks-t) [reply src/trailer.c]: Sorry for not making myself clear, but with "initial design" I referred to upstream git.git, not your code. So I do not complain about your implementation but all the heuristics employed by git.git.

**Comment by brianmario** [comment]
@ethomson how do you feel about this now? :)

**Comment by carlosmn** [comment]
While the upstream code is a good place to start, our style is to provide iterators in order to allow different languages and environments.

If we merge this, we should be looking at making this into an iterator, optionally having the callback version built on top of that.

File: src/trailer.c [hunk 13]
```diff
@@ -0,0 +1,402 @@
...
+}
+
+/* Return the position of the end of the trailers. */
+static int find_trailer_end(const char *buf, size_t len)
+{
+	return len - ignore_non_trailer(buf, len);
+}
+
+static char *find_trailer(const char *message, size_t* len)
+{
```
Comment (carlosmn) [inline src/trailer.c]: The name suggests that we're going to return the offset. Instead we're returning a new buffer and its length. Here we would name that something like `extract_trailer` which would have a `git_buf*` output parameter. `len` here similarly is an output parameter which here we put at the beginning of the parameter list.

File: tests/message/trailer.c [hunk 14]
```diff
@@ -0,0 +1,179 @@
...
+{
+	struct trailer trailers[] = {
+		{"Key", "value"},
+		{NULL, NULL},
+	};
+
+	assert_trailers(
+		"Message\n"
+		"\n"
+		"Key   :   value\n"
```
Comment (carlosmn) [inline tests/message/trailer.c]: Why would this be valid?

File: src/trailer.c [hunk 15]
```diff
@@ -0,0 +1,402 @@
...
+ * is no patch in the message.
+ */
+static int find_patch_start(const char *str)
+{
+	const char *s;
+
+	for (s = str; *s; s = next_line(s)) {
+		if (git__prefixcmp(s, "---") == 0)
+			return s - str;
+	}
```
Comment (carlosmn) [inline src/trailer.c]: Should we be able to use `strstr(str, "\n---")` to find the end of the message?

Comment (carlosmn) [reply src/trailer.c]: I suppose that doesn't work if `---` is at the start of the message because we don't actually have one.

File: src/trailer.c [hunk 16]
```diff
@@ -0,0 +1,402 @@
...
+	return 0;
+}
+
+/*
+ * If the given line is of the form
+ * "<token><optional whitespace><separator>..." or "<separator>...", return the
+ * location of the separator. Otherwise, return -1.  The optional whitespace
+ * is allowed there primarily to allow things like "Bug #43" where <token> is
+ * "Bug" and <separator> is "#".
```
[the remark this answers is not in the document]
Comment (brianmario) [reply src/trailer.c]: The the best of my knowledge, we haven't changed any of the logic from the git-core code. The main thing that it does that we don't (yet) is allow for a few git config options to control how trailers are parsed. Like what the trailer key/value separator is.

**Comment by brianmario** [comment]
Would love one final pass on this. After talking to @carlosmn he convinced me of just using an array-based API as we're likely only working with one or maybe two trailers, for the small percentage of messages they even exist in. Plus we're already OK with having the entire commit message in memory anyway.

He and @charliesome already gave the latest change a look and were cool with things structurally. Based on the prior convo it sounds like this was/is ready to merge once CI is green. I'd love to do that ASAP now that we have a public API figured out that follows existing patterns and stuff.

**Comment by brianmario** [comment]
Another spurious failure on AppVeyor and the travis builds are hung not building the mac builds but this is good to go. Going to merge this and we can iterate on further pulls if need-be.

**Comment by pks-t** [comment]
I'm really surprised to see that being merged by you. It's not customary in libgit2 to merge PRs which are non-trivial by the person who is working on that himself, especially when you are kind of new to the project itself (in terms of having contributed to libgit2, not the libgit2 ecosystem in general). So going forward, I'd kindly ask you to not merge such PRs yourself.

**Comment by ethomson** [comment]
Seconding @pks-t.  I slept on this, which mitigated by fury a little bit.  But I'm _incredibly_ tempted to revert this.  I'm scrolling up and see many comments that haven't been addressed here.

Fundamentally:  **this is not a GitHub project**.  This project is maintained by a community of overworked maintainers who largely do this in their spare time.  If your consumption of this library requires that your code have landed in master, then you either need to wait until we review it, **you fix the issues** or do what the rest of us do and ship a fork.

**Comment by brianmario** [comment]
My sincere apologies on the misstep here. We can revert this if you'd like.

I was under the impression that the feedback that was left unaddressed were things we'd address in further refactorings after initially merging this. Once the API was stable and consistent with the standards in libgit2. I've been working with @carlosmn and @charliesome nearly every day on this.

Be that as it may, the code in question (by my understanding) was essentially copied as-is from git-core. Which we already know to be stable and in-use by the entire git-core ecosystem for years. So any changes being requested were for style and wouldn't affect the functionality of the API were they to happen after merging. @ethomson, your [comment here](https://github.com/libgit2/libgit2/pull/4451#issuecomment-355161727) lead me to believe moving forward with the _existing_ code was ok.

It was never my intention to add anything specific to GitHub here. Git trailer parsing is part of git-core (ala `git-interpret-trailers` and I had the impression that in a perfect world libgit2 would support nearly everything git itself does. So this didn't feel like a one-off API, rather the beginning of support for another git-core feature.

I've been a part of this project both in terms of design, its bindings and its use in production against many billions of requests nearly since the first line of code was written 7 or so years ago. I can assure you I mean the project or the community no harm or have any desire to make it specific to my or the company's needs. I misunderstood that we weren't done with this yet and I'm really sorry for that. I feel terrible :\

**Comment by ethomson** [comment]
> I'm really sorry for that. I feel terrible.

Please don’t.  I didn’t mean to say that I thought you didn’t have the best interests of libgit2 at heart.  I know that you do, and it’s hard to balance competing interests like two open source projects and the one that you actually get paid to work on.

This was a miscommunication and I’m as upset with myself and my inattentiveness.  libgit2 is obviously not in a great place when it comes to resources, and trying to keep a high quality bar and do good code reviews is incredibly hard when you don’t have the time.  Sometimes I don’t realize how much time passes.

That’s no excuse though.  And I’m sorry to have taken my frustration out on you.

Thanks for continuing to contribute.  ❤️
